enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots to new commit
+1. Operating System, greenboot, and MicroShift take actions without any additional intervention
+
+#### Implementation perspective
+
+**First boot into commit with MicroShift**
+
+1. First ostree commit is installed
+1. First ostree commit boots
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
+1. MicroShift startup succeeds
+1. All greenboot checks pass
+1. Greenboot runs green scripts
+   - MicroShift green script sets backup mode to "backup"
+
+**Simple host reboot (cont of previous flow)**
+
+1. First ostree commit shuts down
+1. First ostree commit boots
+1. MicroShift starts up
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift startup succeeds
+1. All greenboot checks pass
+1. Greenboot runs green scripts
+   - MicroShift green script sets backup mode to "backup" (no change)
+
+**Failed upgrade (cont of previous flow)**
+
+1. Second ostree commit is staged
+1. First ostree commit shuts down
+1. Second ostree commit boots
+1. MicroShift starts up
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift startup fails
+1. Greenboot runs red scripts
+    - Set backup mode to "restore"
+1. (Failures may need to repeat to trigger rollback.)
+1. Second ostree commit shuts down
+1. First ostree commit boots
+1. MicroShift starts
+   - Backup mode is "restore",
+   - Backup tool restores the backup compatible with the first ostree commit
+1. MicroShift starts successfully
+1. Greenboot runs green scripts
+   - Set backup mode to "backup"
+
+**Successful upgrade (cont of previous flow)**
+
+1. Third ostree commit is staged
+1. First ostree commit shuts down
+1. Third ostree commit boots
+1. MicroShift starts
+   - Backup mode is "backup",
+   - Backup script runs, creating a backup compatible with the first ostree commit
+1. MicroShift startup succeeds
+1. Green boot runs green scripts
+   - Set backup mode to "backup" (no change)
+
+**Fail first startup, FDO (FIDO Device Onboard) deployment**
+
+1. An ostree commit without MicroShift is installed on the device at the factory.
+1. The device boots at a customer site.
+1. An agent in the ostree commit performs FIDO device onboarding or a similar process to determine the workload.
+1. An ostree commit with MicroShift installed is staged.
+1. The sans-MicroShift commit shuts down.
+1. The with-MicroShift commit starts up.
+1. MicroShift starts
+   - No backup action (did not previously run, so there is nothing to backup, and did not previously fail so no restore needed)
+1. MicroShift startup fails
+1. Greenboot runs red scripts
+   - Set backup mode to "restore"
```

Comment:
This goes back to "first-boot-fails" case I mentioned above. We may want to delete /var/lib/microshift directory completely in this case.

Reply:
This is an open question (also included in the last step of this flow). I'm also partial to removing the directory, but why do you prefer that solution?

Reply:
Because that seems to put the system in a consistent state for the next boot. In this w/f, there will be no MicroShift scripts after the rollback - only /var/lib/microshift directory with stale data will remain.

Reply:
How can we tell first-boot-fails apart from a regular rollback so we only delete in the first-boot-fails case?